_posts/2024-09-20-error-handling.markdown

```diff
@@ -0,0 +1,173 @@
+This means that it takes only 32 bytes of space and requires four instructions to write to.
+Our [`setjmp` implementation](https://github.com/CHERIoT-Platform/cheriot-rtos/blob/e4ebccc60840de78a919c42e6dd7045c99b68245/sdk/include/setjmp.h#L43) is only six instructions as a result (as is longjmp).
+
+# Providing a thread-local unwind handler
+
+The unwinding mechanism described above depends on two things: `setjmp` and somewhere to stash a linked list of `jmp_buf`s.
+The first bit is easy, but the second bit is more complex.
+
+CHERIoT uses the local / global mechanism from CHERI to enforce strong thread isolation.
+Pointers derived from the stack pointer and return addresses may be stored only on the stack.
+This means that at least two of the four registers that must be stored in a `jmp_buf` by `setjmp` can be stored only on the stack.
+
+The traditional approach of keeping the head of the linked list in a global is therefore impossible.
+The slightly more modern approach of using thread-local storage (TLS) would also not work because we do not have thread-local storage.
+
+The obvious solution is therefore to add TLS, but what would that mean?
+Consider a single thread that starts in compartment A, then calls into compartment B, which calls back into A.
+If we have conventional TLS, A could stash a pointer into the part of the stack that it passes to B in TLS and then in the second invocation it would be able to violate compartment isolation (reading and writing all of the on-stack state for B).
+If we built a linked-list of `jmp_buf`s there, then it would be possible for the second invocation of A to jump back to the first, bypassing B and violating the trusted-stack guarantees.
+This would be bad.
+
+What we want is not *thread-local* storage but *compartment-invocation-local* storage.
+Each time a thread enters a new compartment, you should get some storage that is not tied to the depth in the control stack.
+
+To implement this, we looked at the earliest way that operating systems have implemented thread-local storage: reserve some space at the top of the stack.
+
+Now, on entry into a compartment, the switcher will move the stack pointer 16 bytes down before transferring control into the callee.
+Similarly, the loader will reserve 16 bytes at the top of the stack before starting a thread.
+
+This means that you have two pointers worth of space that are easy to find (set the address of the stack pointer to its top, then set the address to eight or 16 bytes below that).
+CHERIoT has a convenient `cgettop` instruction and so this sequence is very short.  First, `cgettop` gives the top address, then `csetaddress` gives a new capability derived from the stack pointer that points to the address.  After that, a `-8` immediate offset to a load or store capability instruction can access the space, so we need only three instructions to load the head of the list.
```

Comment:
By convention this should be multiple lines.  Also hey look, a one-past pointer that we are going to bring back in bounds (as opposed to just a loop terminator or something of that ilk).